src/js_native_api_v8.cc

```diff
@@ -270,6 +270,10 @@ class RefBase : protected Finalizer, RefTracker {
 
  protected:
   inline void Finalize(bool is_env_teardown = false) override {
+    // Force deferring behavior if the finalizer happens to delete this
+    // reference.
+    if (is_env_teardown && RefCount() > 0) _refcount = 0;
```

Comment:
Is this enough on its own ? I'm  wondering about the line below which is:

   if (_delete_self || is_env_teardown) {

given that the comment is that we are trying to force the deferring behaviour, but is_env_teardown being in the if might impact that.

Reply:
I may have missed the case where the finalizer does not explicitly call `napi_delete_reference`, in which case it will defer, but it will never delete – a leak.

Reply:
Actually, it will delete, but I suspect it'll run the finalizer twice, which is also incorrect, because it'll defer, then go back to `~napi_env`'s `FinalizeAll`, which will re-run the finalizer, and call `Delete()` again. I think we need to guard calling the finalizer with `_finalize_ran == false && _finalize_callback != nullptr`.